{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9355d008-3505-4cb1-aaba-82bd04abaa33",
   "metadata": {},
   "outputs": [],
   "source": [
    "from math   import log,floor\n",
    "from random import Random,random as unit_random,choice,shuffle,sample as random_sample\n",
    "\n",
    "\n",
    "# EchyDna--\n",
    "#\tClass to deal with strings of DNA.\n",
    "\n",
    "class EchyDna(object):\n",
    "\tbackground=\"ACGT\"    # default background distribution for generated sequence\n",
    "\n",
    "\tdef __init__(self,lengthOrString,name=None,background=None,seed=None):\n",
    "\t\t\"\"\"\n",
    "\t\tif 1st arg is a length:   generate random sequence of length N\n",
    "\t\tif 1st arg is a string:   use the string as the sequence\n",
    "\t\tif 1st arg is a EchyDna:  use the EchyDna's seq as the sequence\n",
    "\t\tif 1st arg is a list:     concatenate as the sequence; list elements\n",
    "\t\t                          must be strings or EchyDnas\n",
    "\t\tname:                     string to use for name in a few cases (such as\n",
    "\t\t                          .. output to a fasta file)\n",
    "\t\tbackground:               letters to choose from; multiplicities are OK,\n",
    "\t\t                          .. eg \"AAACCGGTTT\"\n",
    "\t\tif seed is None:          use the standard prng, unseeded\n",
    "\t\tif seed is a Random:      use it as the prng\n",
    "\t\tif seed is string or int: use a fresh prng with that seed\n",
    "\t\t\"\"\"\n",
    "\n",
    "\t\tif (name == None): self.name = \"\"\n",
    "\t\telse:              self.name = name\n",
    "\n",
    "\t\t# if the caller has given us a list of sequences, concatenate it\n",
    "\n",
    "\t\tif (isinstance(lengthOrString,list)):\n",
    "\t\t\tseq = []\n",
    "\t\t\tfor item in lengthOrString:\n",
    "\t\t\t\tseq += [str(item)]\n",
    "\t\t\tself.seq = \"\".join(seq)\n",
    "\t\t\treturn\n",
    "\n",
    "\t\t# if the caller has given us a sequence, use it\n",
    "\n",
    "\t\tif (isinstance(lengthOrString,EchyDna)) or (isinstance(lengthOrString,str)):\n",
    "\t\t\tself.seq = str(lengthOrString)\n",
    "\t\t\treturn\n",
    "\n",
    "\t\t# otherwise, generate a random sequence\n",
    "\n",
    "\t\tif (isinstance(lengthOrString,int)):\n",
    "\t\t\tlength = lengthOrString\n",
    "\t\telif (isinstance(lengthOrString,float)):\n",
    "\t\t\tlength = int(floor(lengthOrString))\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"first argument is neither a string nor valid length\")\n",
    "\n",
    "\t\t(_,chooser,_,_) = resolve_prng(seed)\n",
    "\n",
    "\t\tif (length <= 0):\n",
    "\t\t\tself.seq = \"\"\n",
    "\t\telse:\n",
    "\t\t\tif (background == None): background = EchyDna.background\n",
    "\t\t\tself.seq = \"\".join([chooser(background) for _ in range(length)])\n",
    "\n",
    "\t@classmethod\n",
    "\tdef from_fasta(EchyDna,filename,requireFastaHeader=False):\n",
    "\t\t\"\"\"generator; yields each sequence read from the file\"\"\"\n",
    "\t\tf = open(filename,\"rt\")\n",
    "\t\tseqName = None\n",
    "\t\tseqNucs = None\n",
    "\t\tfor line in f:\n",
    "\t\t\tline = line.strip()\n",
    "\t\t\tif (line.startswith(\">\")):\n",
    "\t\t\t\tif (seqName != None):\n",
    "\t\t\t\t\tyield EchyDna(\"\".join(seqNucs),name=seqName)\n",
    "\t\t\t\tseqName = line[1:].strip()\n",
    "\t\t\t\tseqNucs = []\n",
    "\t\t\telif (seqName == None):\n",
    "\t\t\t\tif (requireFastaHeader):\n",
    "\t\t\t\t\traise ValueError(\"The firsrt sequence in \\\"\" + filename + \"\\\" has no fasta header\")\n",
    "\t\t\t\tseqName = \"\"\n",
    "\t\t\t\tseqNucs = [line]\n",
    "\t\t\telse:\n",
    "\t\t\t\tseqNucs += [line]\n",
    "\t\tif (seqName != None):\n",
    "\t\t\tyield EchyDna(\"\".join(seqNucs),name=seqName)\n",
    "\t\tf.close()\n",
    "\n",
    "\tdef fasta(self,filename=None,name=None,wrap=None,append=False,returnIt=True):\n",
    "\t\t\"\"\"\n",
    "\t\tif filename is a string, we'll write a file with that name and close it\n",
    "\t\tif it is a file object, we'll write to it and leave it open\n",
    "\t\tappend is only relevant if filename is a string\n",
    "\t\t\"\"\"\n",
    "\t\tif (name == None): name = self.name\n",
    "\t\tfasta = [\">\" + name]\n",
    "\t\tif (wrap == None) or (wrap == 0):\n",
    "\t\t\tfasta += [self.seq]\n",
    "\t\telse:\n",
    "\t\t\tif (not isinstance(wrap,int)) or (wrap < 0):\n",
    "\t\t\t\traise ValueError(\"wrap argument is not a valid count\")\n",
    "\t\t\tfor ix in range(0,len(self.seq),wrap):\n",
    "\t\t\t\tfasta += [self.seq[ix:ix+wrap]]\n",
    "\t\tfasta = \"\\n\".join(fasta)\n",
    "\t\tif (filename == \"\"): filename = None\n",
    "\t\tif (filename == None):\n",
    "\t\t\treturn fasta\n",
    "\t\tcloseFile = False\n",
    "\t\tif (isinstance(filename,str)):\n",
    "\t\t\tif (append): f = open(filename,\"at\")\n",
    "\t\t\telse:        f = open(filename,\"wt\")\n",
    "\t\t\tcloseFile = True\n",
    "\t\telif (isinstance(filename,file)):\n",
    "\t\t\tf = filename\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"filename argument is not a valid string\")\n",
    "\t\tprint(fasta,file=f)\n",
    "\t\tif (closeFile):\n",
    "\t\t\tf.close()\n",
    "\t\treturn fasta if (returnIt) else None\n",
    "\n",
    "\tdef __str__(self):\n",
    "\t\treturn self.seq\n",
    "\n",
    "\tdef __repr__(self):\n",
    "\t\treturn \"EchyDna(\" + self.seq + \")\"\n",
    "\n",
    "\tdef __len__(self):\n",
    "\t\treturn len(self.seq)\n",
    "\n",
    "\tdef __getitem__(self,slice):\n",
    "\t\treturn EchyDna(self.seq.__getitem__(slice))\n",
    "\n",
    "\tdef __add__(self,other):\n",
    "\t\t\"\"\"a+b concatentates a and b\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn EchyDna(self.seq+str(other))\n",
    "\n",
    "\tdef __iadd__(self,other):\n",
    "\t\t\"\"\"a+=b concatentates b to a\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\tself.seq += str(other)\n",
    "\t\treturn self\n",
    "\n",
    "\tdef __sub__(self,other):\n",
    "\t\t\"\"\"a-b concatentates a and reverse-complement of b\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn EchyDna(self.seq+reverse_complement(str(other)))\n",
    "\n",
    "\tdef __isub__(self,other):\n",
    "\t\t\"\"\"a-=b concatentates reverse-complement of b to a\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\tself.seq += reverse_complement(str(other))\n",
    "\t\treturn self\n",
    "\n",
    "\tdef __mul__(self,other):\n",
    "\t\t\"\"\"a*n concatentates n copies of a; if n is negative the reverse complement of a is used\"\"\"\n",
    "\t\tif (isinstance(other,int)):\n",
    "\t\t\tpass\n",
    "\t\telif (isinstance(other,float)):\n",
    "\t\t\tother = int(floor(other))\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"operand is not a valid count\")\n",
    "\t\tif (other >= 0):\n",
    "\t\t\treturn EchyDna(self.seq * other)\n",
    "\t\telse:\n",
    "\t\t\treturn EchyDna(reverse_complement(self.seq) * -other)\n",
    "\n",
    "\tdef __imul__(self,other):\n",
    "\t\t\"\"\"a*=n replaces a with n copies of itself; if n is negative the reverse complement of a is used\"\"\"\n",
    "\t\tif (isinstance(other,int)):\n",
    "\t\t\tpass\n",
    "\t\telif (isinstance(other,float)):\n",
    "\t\t\tother = int(floor(other))\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"operand is not a valid count\")\n",
    "\t\tif (other >= 0):\n",
    "\t\t\tself.seq *= other\n",
    "\t\telse:\n",
    "\t\t\tself.seq = reverse_complement(self.seq) * -other\n",
    "\t\treturn self\n",
    "\n",
    "\tdef __lshift__(self,other):\n",
    "\t\t\"\"\"a<<n discards n nts on the left (abcde<<1 is bcde)\"\"\"\n",
    "\t\tif (isinstance(other,int)):\n",
    "\t\t\tn = other\n",
    "\t\telif (isinstance(other,float)):\n",
    "\t\t\tn = int(floor(other))\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"operand is not a valid count\")\n",
    "\t\tif (n <= 0): return EchyDna(self.seq)\n",
    "\t\telse:        return EchyDna(self.seq[n:])\n",
    "\n",
    "\tdef __rshift__(self,other):\n",
    "\t\t\"\"\"a>>n discards n nts on the right (abcde>>1 is abcd)\"\"\"\n",
    "\t\tif (isinstance(other,int)):\n",
    "\t\t\tn = other\n",
    "\t\telif (isinstance(other,float)):\n",
    "\t\t\tn = int(floor(other))\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"operand is not a valid count\")\n",
    "\t\tif (n <= 0): return EchyDna(self.seq)\n",
    "\t\telse:        return EchyDna(self.seq[:-n])\n",
    "\n",
    "\tdef __neg__(self):\n",
    "\t\t\"\"\"-a gives reverse-complement of a\"\"\"\n",
    "\t\treturn EchyDna(reverse_complement(self.seq))\n",
    "\n",
    "\tdef __invert__(self):\n",
    "\t\t\"\"\"~a gives un-reversed complement of a\"\"\"\n",
    "\t\treturn EchyDna(forward_complement(self.seq))\n",
    "\n",
    "\tdef __iter__(self):\n",
    "\t\tself.iterIx = 0\n",
    "\t\treturn self\n",
    "\n",
    "\tdef __next__(self):\n",
    "\t\tif (self.iterIx == len(self.seq)):\n",
    "\t\t\traise StopIteration\n",
    "\t\tself.iterIx += 1\n",
    "\t\treturn self.seq[self.iterIx-1]\n",
    "\n",
    "\tdef reverse(self):\n",
    "\t\t\"\"\"a.reverse gives un-complemented reverse of a\"\"\"\n",
    "\t\treturn EchyDna(self.seq[::-1])\n",
    "\n",
    "\tdef upper(self):\n",
    "\t\treturn EchyDna(self.seq.upper())\n",
    "\n",
    "\tdef lower(self):\n",
    "\t\treturn EchyDna(self.seq.lower())\n",
    "\n",
    "\tdef __eq__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq == str(other))\n",
    "\n",
    "\tdef __ne__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq != str(other))\n",
    "\n",
    "\tdef __lt__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq < str(other))\n",
    "\n",
    "\tdef __le__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq <= str(other))\n",
    "\n",
    "\tdef __gt__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq > str(other))\n",
    "\n",
    "\tdef __ge__(self,other):\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (self.seq >= str(other))\n",
    "\n",
    "\tdef __contains__(self,other):\n",
    "\t\t\"\"\"b in a is true iff b is a subtring of a; *is* sensitive to upper/lowercase\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\treturn (str(other) in self.seq)\n",
    "\n",
    "\tdef replace(self,old,new,maxReplace=None):\n",
    "\t\tif (maxReplace != None):\n",
    "\t\t\treturn EchyDna(self.seq.replace(old,new,maxReplace))\n",
    "\t\telse:\n",
    "\t\t\treturn EchyDna(self.seq.replace(old,new))\n",
    "\n",
    "\tdef shuffle(self,seed=None):\n",
    "\t\t(prng,_,_,_) = resolve_prng(seed)\n",
    "\t\tseq = list(self.seq)\n",
    "\t\tif (prng == None): shuffle(seq)\n",
    "\t\telse:              prng.shuffle(seq)\n",
    "\t\treturn EchyDna(\"\".join(seq))\n",
    "\n",
    "\tdef kmers(self,k):\n",
    "\t\tfor ix in range(k,len(self.seq)+1):\n",
    "\t\t\tyield self.seq[ix-k:ix]\n",
    "\n",
    "\tdef mutate(self,subRate=None,indelRate=None,indels=None,background=None,seed=None):\n",
    "\t\t\"\"\"\n",
    "\t\tsubRate:    the per-base probability of a substitution\n",
    "\t\tindelRate:  the per-base probability of an insertion or deletion event\n",
    "\t\tindels:     generally, a function that maps a random number (in 0..1) to\n",
    "\t\t            .. an insertion or deletion length; insertions are positive,\n",
    "\t\t            ..   deletions negative\n",
    "\t\t            .. indels can also be a list of lengths, to be chosen\n",
    "\t\t            ..   with uniform probability\n",
    "\t\t            .. indels can also be an integer N, representing a uniform\n",
    "\t\t            ..   distribution of -N..+N without 0\n",
    "\t\t            .. if this is not provided, indels of length 1 are used\n",
    "\t\tbackground: letters to choose from; multiplicities are OK;\n",
    "\t\tseed:       same as for __init__\n",
    "\t\t\"\"\"\n",
    "\n",
    "\t\t# TODO: incorporate substitute() into this function, to be performed\n",
    "\t\t#       before any of the other mutations\n",
    "\n",
    "\t\t# allow the caller to pass indel information as a pair in either\n",
    "\t\t# indel argument\n",
    "\n",
    "\t\tif (indelRate != None) \\\n",
    "\t\t  and (indels == None) \\\n",
    "\t\t  and (type(indelRate) == tuple) \\\n",
    "\t\t  and (len(indelRate) == 2):\n",
    "\t\t\t(indelRate,indels) = indelRate\n",
    "\t\telif (indels != None) \\\n",
    "\t\t  and (indelRate == None) \\\n",
    "\t\t  and (type(indels) == tuple) \\\n",
    "\t\t  and (len(indels) == 2):\n",
    "\t\t\t(indelRate,indels) = indels\n",
    "\n",
    "\t\t# if the mutation rates are both zero, just return a copy\n",
    "\n",
    "\t\tif (subRate   == None) or (subRate   <= 0): subRate   = 0.0\n",
    "\t\tif (indelRate == None) or (indelRate <= 0): indelRate = 0.0\n",
    "\n",
    "\t\tif (subRate == 0.0) and (indelRate == 0.0):\n",
    "\t\t\treturn EchyDna(self.seq)\n",
    "\n",
    "\t\t# sanity check to make sure the sum of the rates is not more than .99;\n",
    "\t\t# because we will allow back-to-back indels, an indel rate of 1 would\n",
    "\t\t# create an infinite loop\n",
    "\n",
    "\t\tmaxRate   = .99\n",
    "\t\teventRate = subRate + indelRate\n",
    "\n",
    "\t\tif (subRate > maxRate):\n",
    "\t\t\t(subRate,indelRate) = (maxRate,0.0)\n",
    "\t\t\teventRate = maxRate\n",
    "\t\telif (eventRate > maxRate):\n",
    "\t\t\tindelRate = maxRate - subRate\n",
    "\t\t\teventRate = maxRate\n",
    "\n",
    "\t\t# make sure we have an indel length function, if we need one\n",
    "\t\t#  none provided: give 1 bp insert or delete with equal probability;\n",
    "\t\t#  int provided:  choose 1..N and -1..-N equal probabilities\n",
    "\t\t#  list provided: choose from the list with equal probabilities\n",
    "\n",
    "\t\tif (indelRate > 0):\n",
    "\t\t\tif (indels == None):\n",
    "\t\t\t\tindels = lambda u: -1 if (u<0.5) else 1\n",
    "\t\t\telif (type(indels) == int) and (indels != 0):\n",
    "\t\t\t\tmaxLength = abs(indels)\n",
    "\t\t\t\tscale = 2*maxLength - 1e-5\n",
    "\t\t\t\tindels = lambda u: -(1+int(floor(u*scale))) if (u<0.5) \\\n",
    "\t\t\t\t                else 1+int(floor((u-0.5)*scale))\n",
    "\t\t\telif (type(indels) in [list,tuple]) and (len(indels) > 0):\n",
    "\t\t\t\tindelList = indels\n",
    "\t\t\t\tindels = lambda u: indelList[int(floor(u*len(indelList)))]\n",
    "\n",
    "\t\t# build nucleotide substitution tables\n",
    "\n",
    "\t\tif (background == None): background = EchyDna.background\n",
    "\n",
    "\t\tif (subRate != None):\n",
    "\t\t\tntToSubs = {}\n",
    "\t\t\tfor nt in \"ACGT\":\n",
    "\t\t\t\tntToSubs[nt] = [subNt for subNt in background if (subNt != nt)]\n",
    "\t\t\t\tif (len(ntToSubs[nt]) == 0):\n",
    "\t\t\t\t\traise ValueError(\"background contains no substitutions for \\\"%s\\\"\" % nt)\n",
    "\n",
    "\t\t# perform mutation\n",
    "\n",
    "\t\t(_,chooser,spinner,_) = resolve_prng(seed)\n",
    "\n",
    "\t\tseq = []\n",
    "\t\tix = 0\n",
    "\t\twhile (ix < len(self.seq)):\n",
    "\t\t\tnt = self.seq[ix]\n",
    "\n",
    "\t\t\tspin = spinner()\n",
    "\t\t\tif (spin >= eventRate):      # no event\n",
    "\t\t\t\tseq += [nt]\n",
    "\t\t\t\tix  += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\tif (spin >= indelRate):      # substitution event\n",
    "\t\t\t\tif (nt in ntToSubs): seq += [chooser(ntToSubs[nt])]\n",
    "\t\t\t\telse:                seq += [chooser(background)]\n",
    "\t\t\t\tix += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\tspin /= indelRate            # remap spin into 0..1\n",
    "\t\t\tindelLen = indels(spin)\n",
    "\t\t\tif (indelLen == 0):          # no event\n",
    "\t\t\t\tseq += [nt]\n",
    "\t\t\t\tix  += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\tif (indelLen < 0):           # deletion event\n",
    "\t\t\t\tix += (-indelLen)\n",
    "\t\t\telse:                        # insertion event\n",
    "\t\t\t\tseq += [chooser(background) for _ in range(indelLen)]\n",
    "\t\t\t\t# we don't increase ix -- back-to-back indels are allowed\n",
    "\n",
    "\t\t# check for insertion(s) after the end of the sequence\n",
    "\n",
    "\t\tif (indelRate > 0):\n",
    "\t\t\twhile (True):\n",
    "\t\t\t\tspin = spinner()\n",
    "\t\t\t\tif (spin >= indelRate):  # no event\n",
    "\t\t\t\t\tbreak\n",
    "\t\t\t\tspin /= indelRate        # remap spin into 0..1\n",
    "\t\t\t\tindelLen = indels(spin)\n",
    "\t\t\t\tif (indelLen <= 0):      # no event\n",
    "\t\t\t\t\tbreak\n",
    "\t\t\t\tseq += [chooser(background) for _ in range(indelLen)]\n",
    "\n",
    "\t\treturn EchyDna(\"\".join(seq))\n",
    "\n",
    "\tdef substitute(self,numSubs,background=None,seed=None):\n",
    "\t\t\"\"\"\n",
    "\t\tperform a substitution at a specific number of random positions\n",
    "\t\tnumSubs:    number of bases to change\n",
    "\t\tbackground: letters to choose from; multiplicities are OK;\n",
    "\t\tseed:       same as for __init__\n",
    "\t\t\"\"\"\n",
    "\n",
    "\t\t# if there are no substitutions to make, just return a copy\n",
    "\n",
    "\t\tif (numSubs == None) or (numSubs <= 0):\n",
    "\t\t\treturn EchyDna(self.seq)\n",
    "\n",
    "\t\t# clip the number of substitutions to the number of nts in the sequence\n",
    "\n",
    "\t\tvalidPositions = [ix for ix in range(len(self.seq)) if (self.seq[ix] in EchyDna.background)]\n",
    "\t\tnumSubs = min(numSubs,len(validPositions))\n",
    "\n",
    "\t\tif (numSubs <= 0):\n",
    "\t\t\treturn EchyDna(self.seq)\n",
    "\n",
    "\t\t# build nucleotide substitution tables\n",
    "\n",
    "\t\tif (background == None): background = EchyDna.background\n",
    "\n",
    "\t\tntToSubs = {}\n",
    "\t\tfor nt in \"ACGT\":\n",
    "\t\t\tntToSubs[nt] = [subNt for subNt in background if (subNt != nt)]\n",
    "\t\t\tif (len(ntToSubs[nt]) == 0):\n",
    "\t\t\t\traise ValueError(\"background contains no substitutions for \\\"%s\\\"\" % nt)\n",
    "\n",
    "\t\t# perform mutation\n",
    "\n",
    "\t\t(_,chooser,_,sampler) = resolve_prng(seed)\n",
    "\n",
    "\t\tseq = list(self.seq)\n",
    "\t\tfor ix in sampler(validPositions,numSubs):\n",
    "\t\t\tseq[ix] = chooser(ntToSubs[nt])\n",
    "\n",
    "\t\treturn EchyDna(\"\".join(seq))\n",
    "\n",
    "\tdef differences(self,other,symbols=\"-x\"):\n",
    "\t\t\"\"\"a.differences(b) shows the positions where a and b differ\"\"\"\n",
    "\t\tif (not isinstance(other,EchyDna)) and (not isinstance(other,str)):\n",
    "\t\t\traise ValueError(\"operand is not a valid string type\")\n",
    "\t\tother = str(other)\n",
    "\t\tif (len(other) != len(self.seq)):\n",
    "\t\t\traise ValueError(\"operand does not match string length\")\n",
    "\t\tif (not isinstance(symbols,str)) or (len(symbols) != 2):\n",
    "\t\t\traise ValueError(\"symbols argument is not a valid string\")\n",
    "\t\tdiff = lambda a,b: symbols[0] if (a==b) else symbols[1]\n",
    "\t\treturn \"\".join(map(diff,self.seq,other))\n",
    "\n",
    "\n",
    "# geometric_indel--\n",
    "#\tgeometric_indel(p) can be used the indels argument for EchyDna.mutate()\n",
    "\n",
    "def geometric_indel(pExtend):\n",
    "\treturn lambda u: -geometric_distribution(pExtend,2*u) if (u<0.5) \\\n",
    "\t             else geometric_distribution(pExtend,2*u-1)\n",
    "\n",
    "def geometric_distribution(pExtend,u):\n",
    "\treturn int(floor(1+log(1-u)/log(pExtend)))\n",
    "\n",
    "\n",
    "# resolve_prng--\n",
    "#\tFigure out which prng to use, as per the needs of EchyDna.__init__\n",
    "\n",
    "def resolve_prng(seed=None):\n",
    "\t\"\"\"\"\"\"\n",
    "\tprng = None\n",
    "\tif (isinstance(seed,Random)):\n",
    "\t\tprng = seed\n",
    "\t\tchooser = prng.choice\n",
    "\t\tspinner = prng.random\n",
    "\t\tsampler = prng.random_sample\n",
    "\telif (seed != None):\n",
    "\t\tprng = Random()\n",
    "\t\tprng.seed(seed)\n",
    "\t\tchooser = prng.choice\n",
    "\t\tspinner = prng.random\n",
    "\t\tsampler = prng.random_sample\n",
    "\telse:\n",
    "\t\tchooser = choice\n",
    "\t\tspinner = unit_random\n",
    "\t\tsampler = random_sample\n",
    "\treturn (prng,chooser,spinner,sampler)\n",
    "\n",
    "\n",
    "# reverse_complement--\n",
    "\n",
    "complementMap = str.maketrans(\"ACGTSWRYMKBDHVNacgtswrymkbdhvn\",\n",
    "                              \"TGCASWYRKMVHDBNtgcaswyrkmvhdbn\")\n",
    "\n",
    "def reverse_complement(nts):\n",
    "\treturn nts[::-1].translate(complementMap)\n",
    "\n",
    "def forward_complement(nts):\n",
    "\treturn nts.translate(complementMap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "584ba781-b64f-4b90-afb8-c81eecd0023a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from random  import seed as random_seed,randint\n",
    "#import EchyDna\n",
    "EchyDna.background = \"AAACCGGTTT\"\n",
    "random_seed(\"20240327\")\n",
    "\n",
    "flankLength  = 1000\n",
    "itemLengths  = [1000,2000,5000,10000]\n",
    "divergences  = list(range(0,40+1))\n",
    "indelRate    = 0.01 / 2\n",
    "indelLengths = [1,2,3,4,5]\n",
    "\n",
    "isFirstItem = True\n",
    "for d in divergences:\n",
    "\tdivergence = d/100\n",
    "\tfor itemLen in itemLengths:\n",
    "\t\titemName = \"id%d_len%d\" % (100-d,itemLen)\n",
    "\t\tapple  = EchyDna(itemLen)\n",
    "\t\torange = apple.mutate(subRate=divergence,indelRate=indelRate,indels=indelLengths)\n",
    "\t\tif (randint(0,1) == 1): orange = -orange\n",
    "\t\tapple  = EchyDna(flankLength) + apple  + EchyDna(flankLength)\n",
    "\t\torange = EchyDna(flankLength) + orange + EchyDna(flankLength)\n",
    "\t\tapple.fasta (\"../sequences/apple.indels.fa\" ,name=\"APPLE_\"+itemName ,wrap=100,append=not isFirstItem,returnIt=False)\n",
    "\t\torange.fasta(\"../sequences/orange.indels.fa\",name=\"ORANGE_\"+itemName,wrap=100,append=not isFirstItem,returnIt=False)\n",
    "\t\tisFirstItem = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f6e207d-b52b-4484-8a1a-8b762c1f3859",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
